### PR TITLE
Provide support for renaming settings.

### PR DESCRIPTION
**Pull request by dblock**
Signed-off-by: dblock <dblock@amazon.com>

### Description

OpenSearch provides a mechanism for upgrading node settings through upgraders. When an upgrader finds a legacy setting instead of a new one, it replaces it with the new setting and removes the old setting from runtime. I attempted to repurpose those to provide a simple way for plugin authors to rename settings from, for example, `opendistro.jobscheduler.retry_count` to `opensearch.jobscheduler.retry_count`.

When a node is upgraded (or on start of a newer version), upgraders are invoked. Plugins that listen to settings changes will notice. See https://github.com/opensearch-project/job-scheduler/pull/19 for how this is used in the job-scheduler plugin.
 
### Other Ideas

I considered something completely magical that `s/opendistro/opensearch/g` in Settings, but that feels too magical and will inevitably lead to bugs.

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by saratvemulapalli — COMMENTED** [review]
I like the solution, its very simple.
How do we add docs to this feature? (my inner question, how do we let users know this is how you can use it?)
May be few lines of comments like java doc would be great.

File: server/src/main/java/org/opensearch/common/settings/GenericSettingUpgrader.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
```
Comment (dblock) [inline server/src/main/java/org/opensearch/common/settings/GenericSettingUpgrader.java:8]: @saratvemulapalli Wanted to double check that I don't need the "Licensed to Elasticsearch" section in this new file. Correct?

Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/common/settings/GenericSettingUpgrader.java:8]: Correct, the only license header we would need is SPDX for all new files.

Comment (dblock) [reply server/src/main/java/org/opensearch/common/settings/GenericSettingUpgrader.java:8]: And the "Modifications Copyright OpenSearch Contributors." part?

**Comment by dblock** [comment]
> How do we add docs to this feature? (my inner question, how do we let users know this is how you can use it?)

I added JavaDoc to `GenericSettingUpgrader`.

File: server/src/main/java/org/opensearch/common/settings/GenericSettingUpgrader.java (hunk 1 above)
Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/common/settings/GenericSettingUpgrader.java:8]: Yup that is enforced by the license header check and its needed.

**Comment by saratvemulapalli** [comment]
start gradle check

**Comment by dblock** [comment]
@nknize Please help review. While it's not that much work, if you have any suggestions of how we help plugin owners do even less than the diff in https://github.com/opensearch-project/job-scheduler/pull/19/files, I'd love to hear those!

**Comment by Bukhtawar** [comment]
Thanks @dblock ... haven't gone through the code yet but does it work for non dynamic settings too that don't have a settings update consumer

**Comment by dblock** [comment]
> Thanks @dblock ... haven't gone through the code yet but does it work for non dynamic settings too that don't have a settings update consumer

I would say no. This means that for every setting we want to upgrade we'd have to add a dynamic consumer, which may be difficult or easy depending on the case. Do you think this is a big deal?

**Comment by Bukhtawar** [comment]
While I don't think it's a big deal, I do think it's hacky given that dynamic update consumers listen to changes throughout the runtime, not just boot up time which makes it a little non intuitive, also given that it has been forever used along with dynamic settings exclusively.
It'd be great if settings that plugins register, get considered for auto upgrade without needing an explicit plugin subscribed event, at node boot strap. We can check what node boot strap event we could possibly listen onto..

**Comment by dblock** [comment]
> While I don't think it's a big deal, I do think it's hacky given that dynamic update consumers listen to changes throughout the runtime, not just boot up time which makes it a little non intuitive, also given that it has been forever used along with dynamic settings exclusively.
> It'd be great if settings that plugins register, get considered for auto upgrade without needing an explicit plugin subscribed event, at node boot strap. We can check what node boot strap event we could possibly listen onto..

I see what you mean now, but I think I didn't explain it well enough. 

Consider setting OLD that is declared in a plugin and has its value set. When the node joins a cluster, the value of OLD is read and used in the plugin code.

Next version of plugin. Now we want to upgrade OLD into NEW. We implement an upgrader for OLD, which turns it into NEW. Our new code reads the value of NEW only, and uses that. When a node joins a cluster, OLD is read, then NEW is read but is not set, and its default value is used. So now we are using the default value of NEW, which is not what we want, cause we want the upgrade! After this initial load upgraders run, they remove OLD and upgrade it into NEW, the value of NEW changes, and because we have a listener we get to use this value, and all is well.

If we don't listen, the other option is to something like "if NEW is not set, try OLD, otherwise use default value of NEW". I'll take a look at it, but maybe it's better to support changing settings at runtime? Do you have an example where it would be a bad idea?

**Comment by Bukhtawar** [comment]
Have you looked at this part which probably is done at node boot strap. I was wondering if plugin upgrade registered settings gets picked up here and "if NEW is not set, try OLD, otherwise use default value of NEW"

https://github.com/opensearch-project/OpenSearch/blob/0ba0e7cc26060f964fcbf6ee45bae53b3a9941d0/server/src/main/java/org/opensearch/gateway/GatewayMetaState.java#L228-L238

**Comment by dblock** [comment]
@Bukhtawar Sort of.

After a bit more code reading I realized that there's actually a fallback mechanism for settings. If a setting isn't set, it can fallback to another one. Working example in https://github.com/opensearch-project/job-scheduler/pull/20 in which I renamed opendistro settings to legacy settings, marked them as deprecated, and fall back to them from opensearch settings. This does work.

Interestingly it's the same load order as with upgraders. On start the settings are defaulted, then when the node joins a cluster the node settings are loaded and used. 

- If you put settings in opensearch.yml, then you get fallback right away. Thus https://github.com/opensearch-project/job-scheduler/pull/20 will never see the "wrong" setting, but https://github.com/opensearch-project/job-scheduler/pull/19 would.
- If you set cluster settings, the node will get them when it joins the cluster. Thus https://github.com/opensearch-project/job-scheduler/pull/20 and https://github.com/opensearch-project/job-scheduler/pull/19 have the same effect.

**Comment by dblock** [comment]
I think I'd prefer to merge #665 than this one, will close for now.

**Pull request closed without merging**